### DISCUSSION TITLE
Sanity CMS with Next-intl

### DISCUSSION THREAD

**Discussion by donhamalainen** [amannn/next-intl#458] [General]
Atm I think can & how I do the Sanity Localization and Next-intl integration. Have any one experience?

**Comment by jangir-ritik** [comment]
I am trying to achieve the same now. Tried using Strapi as the CMS but the developer experience had been horrible. How was your experience? Did you do it?

**Comment by markoleavy** [comment]
I've recently integrated next-intl on a Sanity based project.
I used [@sanity/document-internationalization](https://www.sanity.io/plugins/document-internationalization), then used the `Locales` values from next-intl as  ids for  `supportedLanguages` in `sanity.config`. Then added this filter to each doc query (in graphql): `{ language: { eq: ${ locale } } }`
```js
documentInternationalization({
      supportedLanguages: [
        {id: 'it', title: 'Italiano'},
        {id: 'en-US', title: 'Us English'},
      ],
      schemaTypes: ...translatedTypes,
      weakReferences: true,
    }),
```

**Reply by jangir-ritik** [reply]
Hi @markoleavy, I'm still deciding between document-based and field-based translation methods for my project. [link to sanity translation docs](https://www.sanity.io/docs/localization#bab220d4df60) Based on my understanding, I think I need to use field-based translations because the keys remain constant while the content varies. [sanity internationalized array - why-store-localized-field-data-like-this](https://github.com/sanity-io/sanity-plugin-internationalized-array?tab=readme-ov-file#why-store-localized-field-data-like-this).
I have 7 language dictionaries stored in the messages directory, each in its own JSON file.

However, I'm concerned that using field-based translations might exceed Sanity's attribute limit for the free tier. [sanity attribute limit](https://www.sanity.io/docs/attribute-limit)

What do you suggest?

```
├── messages
│   ├── ar.json
│   ├── en.json
│   ├── es.json
│   ├── fr.json
│   ├── hi.json
│   ├── mr.json
│   └── sw.json
├── src
│   ├── app
│   │   └── [locale]
│   │       ├── ...
│   ├── i18n.ts
│   ├── middleware.ts
│   ├── navigation.ts
│   └── pages
│       └── api
│           └── send-mail.js
├── tailwind.config.js
├── tsconfig.json
└── ...
```

The top level structure of content dictionary is 
`[
  "Homepage",
  "Footer",
  "HamburgerMenu",
  "XOne",
  "XStudio",
  "ContactUs",
  "Terms",
  "FAQ",
  "XWallet"
]`

Each page is further divided into sections  
```

  "Homepage": {
    "HeroSection": {
      "title": {
        "beforeSpan": "",
        "spanContent": "",
        "afterSpan": ""
      },
      "linkButton": {
        "isExternal": false,
        "url": "",
        "title": ""
      }
    },
    "DigitalAgeSection": {
      "description": "",
      "title": {
        "beforeSpan": "",
        "spanContent": "",
        "afterSpan": ""
      },
      "linkButton": {
        "isExternal": false,
        "url": "",
        "title": ""
      },
      "super": ""
    },
    "VerifiableCredentialsSection": {
      "title": {
        "beforeSpan": "",
        "spanContent": "",
        "afterSpan": ""
      },
      "super": ""
    },
    "ValuesSection": {
      "title": "",
      "list": [
        {
          "id": 0,
          "title": "",
          "description": ""
        }
      ]
    },
    "TrustSection": {
      "title": "",
      "description": "",
      "linkButton": {
        "url": "",
        "title": ""
      }
    },
    "SolveSection": {
      "super": "",
      "title": "",
      "linkButton": {
        "isExternal": false,
        "url": "",
        "title": ""
      },
      "list": [
        ""
      ]
    },
    "DitchSection": {
      "super": "",
      "title": "",
      "linkButton": {
        "isExternal": false,
        "url": "",
        "title": ""
      },
      "list": [
        ""
      ]
    },
    "BeingSection": {
      "super": "",
      "title": "",
      "linkButton": {
        "isExternal": false,
        "url": "",
        "title": ""
      },
      "list": [
        ""
      ]
    },
    "IndustriesSection": {
      "title": "",
      "list": [
        {
          "id": "",
          "title": "",
          "description": ""
        }
      ]
    },
    "UnlockSection": {
      "title": "",
      "linkButton": {
        "isExternal": false,
        "title": "",
        "url": ""
      }
    }
  },
```

**Reply by markoleavy** [reply]
hi @jangir-ritik. I used document-based translation, as my project uses the Sanity GraphQL endpoint, with which you can't filter documents (nor arrays contents) based on values of array elements. Using document-based translation I can get a docs/pages contents filtering the locale/language field. 
Apart from the GraphQL issue, I think it's a very good way to keep things in order for editors.

> I think I need to use field-based translations because the keys remain constant while the content varies.

I don't understand your need to keep keys constant, can you explain it?

**Comment by jangir-ritik** [comment]
hello again, 
Consider this 
   ```
 "HeroSection": {
      "title": {
        "beforeSpan": "Step into the Future of ",
        "spanContent": "Digital Credential",
        "afterSpan": " Management with "
      },
      "linkButton": {
        "isExternal": false,
        "url": "/contact",
        "title": "Know More"
      }
    },
```

and 

```
    "HeroSection": {
      "title": {
        "beforeSpan": "विश्वास बनाएं ",
        "spanContent": "सत्यापन करने योग्य प्रमाणपत्रों",
        "afterSpan": " के साथ डिजिटल भविष्य के लिए"
      },
      "linkButton": {
        "isExternal": false,
        "url": "/संपर्क",
        "title": "संपर्क में रहें"
      }
    },
```

what I meant by keys remaining the same is that beforeSpan, spanContent, afterSpan, isExternal and other keys will remain the same but the content inside of it would change based on the locale

**Reply by markoleavy** [reply]
Mmmhhh, still a bit confused. Are you trying to manage the 'message' translation sheet through Sanity?
I use Sanity to translate any editable content, and next-intl messages to translate js hardcoded words or sentences and that's it.
